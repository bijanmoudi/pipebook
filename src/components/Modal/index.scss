/*====================================
=            Modal Styles            =
====================================*/

@import '../../assets/styles/constants/layout';
@import '../../assets/styles/helpers/colors';
@import '../../assets/styles/helpers/units';

.modal {
	z-index: 100;
	position: fixed;
	top: 0;
	bottom: 0;
	left: 0;
	right: 0;
	display: flex;
	align-items: center;
	justify-content: center;
	background-color: get-color('black', 0.4);
	cursor: pointer;
	&__inner {
		display: flex;
		flex-direction: column;
		width: rem(430);
		max-width: 100%;
		max-height: 100%;
		overflow: hidden;
		opacity: 1;
		will-change: transform, opacity;
		transform: scale(1);
		transition: transform 250ms ease, opacity 250ms ease;
		background-color: get-color('white');
		box-shadow: 0 0 rem(20) get-color('black', 0.3);
		cursor: initial;
	}
	&__header,
	&__content,
	&__footer {
		padding-left: rem(20);
		padding-right: rem(20);
	}
	&__header,
	&__footer {
		display: flex;
		align-items: center;
		justify-content: space-between;
		min-height: rem(58);
		background-color: get-color('gray-light-ultra');
	}
	&__header {
		padding-top: rem(14);
		padding-bottom: rem(14);
		border-bottom: rem(2) solid get-color('gray-light-very');
		line-height: 1;
		& > * {
			padding-top: rem(5);
			padding-bottom: rem(5);
		}
	}
	&__content {
		flex-grow: 1;
		padding-top: rem(35);
		padding-bottom: rem(35);
		overflow: auto;
		p {
			font-size: rem(14);
		}
	}
	&__footer {
		flex-direction: row-reverse;
		padding-top: rem(10);
		padding-bottom: rem(10);
		border-top: rem(2) solid get-color('gray-light-very');
		box-shadow: 0 rem(-2) rem(2) get-color('black', 0.015);
	}
	&__close {
		display: inline-block;
		padding-left: rem(10);
		padding-right: rem(10);
		border: 0 none;
		background: none;
		color: get-color('gray-dark');
		line-height: 1;
		font-size: em(27);
		font-weight: $semi-bold-font-weight;
		line-height: em(14, 27);
		overflow: hidden;
		&:before {
			display: block;
			content: 'Ã—';
			line-height: inherit;
		}
		&:hover,
		&:focus {
			color: get-color('gray-dark-extra');
		}
	}
	&--is-loading {
		.modal__inner {
			opacity: 0;
			transform: scale(0);
		}
	}
}
