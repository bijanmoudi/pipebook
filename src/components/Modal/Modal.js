import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Button } from '../../components';
import { randomString } from '../../utils';
import './index.scss';

const Modal = ({
	id,
	label,
	title,
	content,
	pageTitle,
	hasCloseButton,
	closeButtonContent,
	closeButtonTitle,
	extraFooterContent,
	closeByOverlay,
	open,
	loading,
	onOpen,
	onClose,
	className,
	...restProps
}) => {
	id = id || `modal--${randomString()}`;
	const [isOpen, setOpen] = useState(open);
	const handleEscape = e => {
		e.key === 'Escape' && handleClose();
	};
	const handleClose = () => {
		setOpen(false);
		document.title.replace(` - ${pageTitle}`, '');
		onClose();
	};
	useEffect(() => {
		if (isOpen && !loading) {
			setTimeout(
				() => (document.title += pageTitle ? ` - ${pageTitle}` : '')
			);
			onOpen();
		}
		// eslint-disable-next-line
	}, [isOpen, loading]);
	// eslint-disable-next-line
	useEffect(() => {
		const previousActiveElement = document.activeElement;
		setOpen(open);
		window.addEventListener('keyup', handleEscape);
		if (open) {
			document.documentElement.style.overflow = 'hidden';
			document.getElementsByTagName('body')[0].style.overflowY = 'scroll';
		}
		return () => {
			window.removeEventListener('keyup', handleEscape);
			document.getElementsByTagName('body')[0].style.overflow = 'hidden';
			document.documentElement.style.overflowY = 'scroll';
			previousActiveElement.focus();
		};
	});
	return isOpen ? (
		<div
			id={id}
			role="dialog"
			aria-modal="true"
			aria-labelledby={`${id}__label`}
			onClick={e => {
				e.preventDefault();
				e.target === e.currentTarget && closeByOverlay && handleClose();
			}}
			className={classnames(
				'modal',
				{ [className]: className },
				{ 'modal--is-loading': loading }
			)}
			{...restProps}
		>
			{label || title ? (
				<span
					id={`${id}__label`}
					className="visuallyhidden"
					aria-live="polite"
				>
					{label || title}
				</span>
			) : null}
			<div className="modal__inner">
				<header className="modal__header">
					<h2 className="modal__title">{title}</h2>
					<button
						className="modal__close"
						title="Close this modal"
						onClick={handleClose}
					>
						<span className="visuallyhidden">Close</span>
					</button>
				</header>
				{content ? (
					<section role="main" className="modal__content">
						{content}
					</section>
				) : null}
				{hasCloseButton || extraFooterContent ? (
					<footer className="modal__footer">
						{extraFooterContent}
						{hasCloseButton ? (
							<Button>
								<button
									title={
										closeButtonTitle || 'Back to the list'
									}
									onClick={handleClose}
								>
									{closeButtonContent || (
										<>
											Back
											<span className="visuallyhidden">
												{' '}
												to the list
											</span>
										</>
									)}
								</button>
							</Button>
						) : null}
					</footer>
				) : null}
			</div>
		</div>
	) : null;
};

Modal.propTypes = {
	id: PropTypes.oneOfType([() => null, PropTypes.string]),
	label: PropTypes.oneOfType([() => null, PropTypes.string]),
	title: PropTypes.oneOfType([() => null, PropTypes.string]),
	content: PropTypes.oneOfType([() => null, PropTypes.string]),
	pageTitle: PropTypes.oneOfType([() => null, PropTypes.string]),
	hasCloseButton: PropTypes.bool,
	closeButtonContent: PropTypes.oneOfType([() => null, PropTypes.element]),
	closeButtonTitle: PropTypes.oneOfType([() => null, PropTypes.string]),
	extraFooterContent: PropTypes.oneOfType([() => null, PropTypes.element]),
	closeByOverlay: PropTypes.bool,
	open: PropTypes.bool,
	loading: PropTypes.bool,
	onOpen: PropTypes.func,
	onClose: PropTypes.func
};

Modal.defaultProps = {
	id: null,
	label: null,
	title: null,
	content: null,
	pageTitle: null,
	hasCloseButton: true,
	closeButtonContent: null,
	closeButtonTitle: null,
	extraFooterContent: null,
	closeByOverlay: true,
	open: false,
	loading: false,
	onOpen: () => {},
	onClose: () => {}
};

export default Modal;
