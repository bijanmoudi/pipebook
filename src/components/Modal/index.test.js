import React from 'react';
import renderer from 'react-test-renderer';
import Modal from './Modal';

const props = {
	id: 'new-person-modal',
	label: 'Add new person',
	title: 'Add new person',
	content: <input type="text" />,
	hasCloseButton: false,
	closeButtonContent: 'Close',
	closeButtonTitle: 'Close this modal',
	extraFooterContent: (
		<p>
			<span class="color-red">(*)</span> Required fields
		</p>
	),
	closeByOverlay: false,
	open: true,
	loading: true
};

describe('Testing "Modal" component', () => {
	it('Rendering closed modal', () => {
		const tree = renderer.create(<Modal id={props.id} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering loading modal', () => {
		const tree = renderer
			.create(<Modal id={props.id} open={props.open} loading={props.loading} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering with customized footer and no title', () => {
		const tree = renderer
			.create(
				<Modal
					id={props.id}
					open={props.open}
					content={props.content}
					closeButtonContent={props.closeButtonContent}
					closeButtonTitle={props.closeButtonTitle}
					extraFooterContent={props.extraFooterContent}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering with all props provided', () => {
		const tree = renderer
			.create(
				<Modal
					id={props.id}
					label={props.label}
					title={props.title}
					content={props.content}
					hasCloseButton={props.hasCloseButton}
					closeButtonContent={props.closeButtonContent}
					closeButtonTitle={props.closeButtonTitle}
					extraFooterContent={props.extraFooterContent}
					closeByOverlay={props.closeByOverlay}
					open={props.open}
					loading={props.loading}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
