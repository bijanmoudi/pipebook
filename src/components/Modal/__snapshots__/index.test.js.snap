// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing "Modal" component Rendering closed modal 1`] = `null`;

exports[`Testing "Modal" component Rendering loading modal 1`] = `
<div
  aria-labelledby="new-person-modal__label"
  aria-modal="true"
  className="modal modal--is-loading"
  id="new-person-modal"
  onClick={[Function]}
  role="dialog"
>
  <div
    className="modal__inner"
  >
    <header
      className="modal__header"
    >
      <h2
        className="modal__title"
      />
      <button
        className="modal__close"
        onClick={[Function]}
        title="Close this modal"
      >
        <span
          className="visuallyhidden"
        >
          Close
        </span>
      </button>
    </header>
    <footer
      className="modal__footer"
    >
      <button
        className="button"
        onClick={[Function]}
        title="Back to the list"
      >
        Back
        <span
          className="visuallyhidden"
        >
           
          to the list
        </span>
      </button>
    </footer>
  </div>
</div>
`;

exports[`Testing "Modal" component Rendering with all props provided 1`] = `
<div
  aria-labelledby="new-person-modal__label"
  aria-modal="true"
  className="modal modal--is-loading"
  id="new-person-modal"
  onClick={[Function]}
  role="dialog"
>
  <span
    aria-live="polite"
    className="visuallyhidden"
    id="new-person-modal__label"
  >
    Add new person
  </span>
  <div
    className="modal__inner"
  >
    <header
      className="modal__header"
    >
      <h2
        className="modal__title"
      >
        Add new person
      </h2>
      <button
        className="modal__close"
        onClick={[Function]}
        title="Close this modal"
      >
        <span
          className="visuallyhidden"
        >
          Close
        </span>
      </button>
    </header>
    <section
      className="modal__content"
      role="main"
    >
      <input
        type="text"
      />
    </section>
    <footer
      className="modal__footer"
    >
      <p>
        <span
          class="color-red"
        >
          (*)
        </span>
         Required fields
      </p>
    </footer>
  </div>
</div>
`;

exports[`Testing "Modal" component Rendering with customized footer and no title 1`] = `
<div
  aria-labelledby="new-person-modal__label"
  aria-modal="true"
  className="modal"
  id="new-person-modal"
  onClick={[Function]}
  role="dialog"
>
  <div
    className="modal__inner"
  >
    <header
      className="modal__header"
    >
      <h2
        className="modal__title"
      />
      <button
        className="modal__close"
        onClick={[Function]}
        title="Close this modal"
      >
        <span
          className="visuallyhidden"
        >
          Close
        </span>
      </button>
    </header>
    <section
      className="modal__content"
      role="main"
    >
      <input
        type="text"
      />
    </section>
    <footer
      className="modal__footer"
    >
      <p>
        <span
          class="color-red"
        >
          (*)
        </span>
         Required fields
      </p>
      <button
        className="button"
        onClick={[Function]}
        title="Close this modal"
      >
        Close
      </button>
    </footer>
  </div>
</div>
`;
